ecs-cli/modules/clients/aws/route53/servicediscovery_client.go

```diff
@@ -0,0 +1,135 @@
+// Copyright 2015-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+// Package route53 contains functions for working with the route53 APIs
+// that back ECS Service Discovery
+package route53
+
+import (
+	"github.com/aws/amazon-ecs-cli/ecs-cli/modules/clients"
+	"github.com/aws/amazon-ecs-cli/ecs-cli/modules/config"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/route53"
+	"github.com/aws/aws-sdk-go/service/servicediscovery"
+)
+
+// FindPrivateNamespace returns the ID(s) of the private namespace with the given name and vpc
+func FindPrivateNamespace(name, vpc string, config *config.CommandConfig) (*string, error) {
+	region := aws.StringValue(config.Session.Config.Region)
+	var nameMatch []*string
+	r53Client := newRoute53Client(config)
+	sdClient := newSDClient(config)
+
+	err := listNamespaces(false, sdClient, func(namespace *servicediscovery.NamespaceSummary) bool {
+		if aws.StringValue(namespace.Name) == name {
+			nameMatch = append(nameMatch, namespace.Id)
+		}
+		return true
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	for _, namespaceID := range nameMatch {
+		hasVPC, err := checkNamespaceVPC(namespaceID, vpc, region, r53Client, sdClient)
+		if err != nil {
+			return nil, err
+		}
+		if hasVPC {
+			return namespaceID, nil
+		}
+	}
+
+	return nil, nil
```

Comment:
retuning nil for both the return value and the error seems like an anti-pattern. when would this condition be true?

Reply:
Returning `nil, nil` means that no matching namespace was found, and that there were no API errors. Not finding a matching namespace and experiencing an error are distinct cases which I want to differentiate.

Reply:
Gotcha, is there value in returning an additional bool to indicate failure in finding a namespace?

Reply:
I don't think there's a need for that. This is roughly how the code that consumes this package will look:

```
namespaceID, err := FindPrivateNamespace(name, vpc, config)
if err != nil {
    return err // terminate execution and log the error to the user
}

if namespaceID != nil {
    // use the existing namespace
} else {
    // namespace doesn't exist, we need to create one
}
```